File: src/double_linked_list/double_linked_list.c
/**
 * Author: José Emiliano Cabrera Blancas (jemiliano.cabrera@gmail.com)
 *
 */

#include "double_linked_list/double_linked_list.h"

#include "types/types.h"
#include "half_edge/half_edge.h"
#include "face/face.h"

#include <stdlib.h>
#include <stdio.h>

list* init_double_linked_list(item_type type)
{
	struct double_linked_list* list;
	list = (struct double_linked_list*) malloc(sizeof(struct double_linked_list));
	

	if (list == NULL) {
		printf("Ya no hay memoria disponible: init_double_linked_list()\n");
		exit(EXIT_FAILURE);
	}

	list->size = 0;
	list->head = list->tail = NULL;
	list->type = type;

	return list;
}

void destroy_double_linked_list(list* list)
{
	if (list == NULL)
		return;
		
	while(!empty_list(list))
		pop_front(list);

	free(list);
}

int empty_list(list* list)
{
	if (list == NULL) {
		return TRUE;
	} else if (list->size == 0) {
	 	return TRUE;
	} else {
		return FALSE;
	}
}

item* init_list_item()
{
	struct list_item* item;
	item = (struct list_item*) malloc(sizeof(struct list_item));
	
	if (item == NULL) {
		printf("Ya no hay memoria disponible: init_list_item()");
		exit(EXIT_FAILURE);
	}

	item->element = NULL;
	item->left = NULL;
	item->right = NULL;
	
	return item;
}

void destroy_list_item(item* item)
{
	free(item);
}

list* create_copy_list(list* list)
{
	struct double_linked_list* copy;
	copy = init_double_linked_list(list->type);
	
	struct list_item* tmp;
	tmp = list->head;
	
	while(tmp != NULL) {
		push_back(copy, tmp->element);
		tmp = tmp->right;
	}

	return copy;
}

void push_back(list* list,  void* element)
{
	struct list_item *tmp1, *tmp2;

	tmp2 = init_list_item();
	tmp2->element = element;

	
	if (empty_list(list)) {
		list->head = list->tail = tmp2;
		list->size += 1;
		return;
	}
	
	tmp1 = list->tail;
	tmp1->right = tmp2;
	
	tmp2->left = tmp1;
  	list->tail = tmp2;
	list->size += 1;
	return;
}

void push_front(list* list, void* element)
{
	struct list_item *tmp1, *tmp2;

	tmp2 = init_list_item();

	tmp2->element = element;

	
	if (empty_list(list)) {
		list->head = list->tail = tmp2;
		list->size += 1;
		return;
	}
	
	tmp1 = list->head;
	tmp1->left = tmp2;
	
	tmp2->right = tmp1;
	list->head = tmp2;

	list->size += 1;
	return;
}

void* pop_back(list* list)
{
	
	if (empty_list(list))
		return NULL;

	struct list_item *tmp1, *tmp2;
	tmp1 = list->tail;
	tmp2 = tmp1->left;
	
	void* element;
	
	list->size -= 1;
	if (list->size == 0) {
		list->head = list->tail = NULL;
	} else {
		tmp2->right = NULL;
		list->tail = tmp2;
	}
	
	element = tmp1->element;
	destroy_list_item(tmp1);
	
	return element;
}

void* pop_front(list* list)
{
	if (empty_list(list)) 
		return NULL;

	struct list_item *tmp1, *tmp2;
	tmp1 = list->head;	
	tmp2 = tmp1->right;
	
	void* element;
	
	list->size -= 1;
	if (list->size == 0) {
		list->head = list->tail = NULL;
	} else {
		tmp2->left = NULL;
		list->head = tmp2;
	}

	element = tmp1->element;

	destroy_list_item(tmp1);
	return element;
}

void* pick_back(list* list)
{
	if (empty_list(list))
		return NULL;
	else
		return list->tail->element;
}

void* pick_front(list* list)
{
	if (empty_list(list))
		return NULL;
	else
		return list->head->element;
}

File: src/shared_libraries/double_linked_list.rb
# -*- coding: utf-8 -*-
# 
# Author: José Emiliano Cabrera Blancas (jemiliano.cabrera@gmail.com)
#
require 'ffi'

require File.expand_path(File.join(File.dirname(__FILE__),"points"))
require File.expand_path(File.join(File.dirname(__FILE__),"half_edge"))

#
# Modulo para mapear la lista doblemente ligada y sus funciones de C a Ruby
#
module List
  extend FFI::Library
  ffi_lib "lib/libdouble_linked_list.so"

  Item_type = enum(:point, 1,
                   :half_edge,
                   :face,
                   :x)

  class List < FFI::ManagedStruct
    layout :head, :pointer,
           :tail, :pointer,
           :size, :int,
           :element, Item_type

    def self.release(ptr)
    List.free_object(ptr)
    end
    
  end
  
  attach_function :init_double_linked_list, [Item_type], List.by_ref
  attach_function :destroy_double_linked_list, [List.by_ref], :void
  
  attach_function :push_back, [List.by_ref, :pointer], :void
  attach_function :push_front, [List.by_ref, :pointer], :void

  attach_function :empty_list, [List.by_ref], :int

  attach_function :pop_back, [List.by_ref], :pointer
  attach_function :pop_front, [List.by_ref], :pointer
  
  attach_function :pick_back, [List.by_ref], :pointer
  attach_function :pick_front, [List.by_ref], :pointer

  attach_function :create_copy_list, [List.by_ref], List.by_ref

  # Public: Convierte una lista de puntos de intersecicon 
  # de C a un arreglo de Ruby.
  # 
  # list - La lista que se quiere convertir a un arreglo.
  # 
  # Examples
  #    
  #     List.to_a((1,2),(2,4)) Recuerda que la entrada es una lista de C.
  #     # => [[1,2],[2,4]]
  #
  # Regresa un arreglo
  def self.to_a(list)
    array = Array.new
    while list[:size] != 0   
      point = Points.to_a(cast_point(pop_front(list)))

      point[2] = cast_half_edge(point[2])
      point[3] = cast_half_edge(point[3])

      point[2] = [[point[2][:first][:x],
                   point[2][:first][:y]],
                  [point[2][:last][:x],
                   point[2][:last][:y]]]
      
      point[3] = [[point[3][:first][:x],
                   point[3][:first][:y]],
                  [point[3][:last][:x],
                   point[3][:last][:y]]]
      
      array << point
    end
    array
  end  
end
